File: fiesta/src/polygon.f90

subroutine triangle_area(xa, ya, xb, yb, xc, yc, area)

  ! Determines the area of a triangle.
  !
  ! Parameters
  ! ----------
  ! xa : float
  !   X coordinate of point a.
  ! ya : float
  !   Y coordinate of point a.
  ! xb : float
  !   X coordinate of point b.
  ! yb : float
  !   Y coordinate of point b.
  ! xc : float
  !   X coordinate of point c.
  ! yc : float
  !   Y coordinate of point c.
  !
  ! Returns
  ! -------
  ! area : float
  !   Area of the triangle.

  implicit none
  integer, parameter :: dp = kind(1.d0)

  ! define variables

  real(kind=dp), intent(in) :: xa, ya, xb, yb, xc, yc
  real(kind=dp), intent(out) :: area

  ! Calculate area.

  area = 0.5*abs(xa*(yb-yc) + xb*(yc - ya) + xc*(ya - yb))

end subroutine triangle_area


subroutine sum_triangle_area(xas, yas, xbs, ybs, xcs, ycs, ntri, area)

  ! Determines the area of a triangle.
  !
  ! Parameters
  ! ----------
  ! xas : array
  !   X coordinates of points a.
  ! yas : array
  !   Y coordinates of points a.
  ! xbs : array
  !   X coordinates of points b.
  ! ybs : array
  !   Y coordinates of points b.
  ! xcs : array
  !   X coordinates of points c.
  ! ycs : array
  !   Y coordinates of points c.
  ! ntri : int
  !   Number of triangles
  !
  ! Returns
  ! -------
  ! area : float
  !   Total area of the triangles.

  implicit none
  integer, parameter :: dp = kind(1.d0)

  ! define variables

  integer, intent(in) :: ntri
  real(kind=dp), intent(in) :: xas(ntri), yas(ntri), xbs(ntri), ybs(ntri), xcs(ntri), ycs(ntri)
  real(kind=dp), intent(out) :: area

  integer :: i
  real(kind=dp) :: area_tri

  ! loop over triangles and calculate the areas.

  area = 0.

  do i = 1, ntri
    call triangle_area(xas(i), yas(i), xbs(i), ybs(i), xcs(i), ycs(i), area_tri)
    area = area + area_tri
  end do

end subroutine sum_triangle_area

File: fiesta/src/polygon.py
import numpy as np
from numba import njit


@njit
def triangle_area(xa: float, ya: float, xb: float, yb: float, xc: float, yc: float) -> float:
    """
    Determines the area of a triangle given its vertex coordinates.

    Parameters
    ----------
    xa : float
        X-axis coordinate for point A of the triangle.
    ya : float
        Y-axis coordinate for point A of the triangle.
    xb : float
        X-axis coordinate for point B of the triangle.
    yb : float
        Y-axis coordinate for point B of the triangle.
    xc : float
        X-axis coordinate for point C of the triangle.
    yc : float
        Y-axis coordinate for point C of the triangle.
    
    Yields
    ------
    Area of the triangle.
    """
    return 0.5 * abs(xa * (yb - yc) + xb * (yc - ya) + xc * (ya - yb))


@njit
def sum_triangle_area(
    xas: np.ndarray, yas: np.ndarray, xbs: np.ndarray, ybs: np.ndarray, xcs: np.ndarray, ycs: np.ndarray
) -> np.ndarray:
    """
    Computes the total area of multiple triangles.

    Parameters
    ----------
    xas : array
        X-axis coordinates for point A of the triangles.
    yas : array
        Y-axis coordinates for point A of the triangles.
    xbs : array
        X-axis coordinates for point B of the triangles.
    ybs : array
        Y-axis coordinates for point B of the triangles.
    xcs : array
        X-axis coordinates for point C of the triangles.
    ycs : array
        Y-axis coordinates for point C of the triangles.
    
    Returns
    -------
    total_area : array
        Area for each triangle.
    """
    ntri = len(xas)
    total_area = 0.0
    
    for i in range(ntri):
        total_area += triangle_area(xas[i], yas[i], xbs[i], ybs[i], xcs[i], ycs[i])
    
    return total_area
